plugins/by-name/typescript-tools/default.nix

```diff
@@ -157,38 +157,68 @@ in
       };
     };
   };
-  config = mkIf cfg.enable {
-    extraPlugins = [ cfg.package ];
-
-    plugins.lsp.postConfig =
-      with cfg;
-      let
-        options = {
-          inherit handlers;
-          on_attach = onAttach;
-
-          settings = with settings; {
-            separate_diagnostic_server = separateDiagnosticServer;
-            publish_diagnostic_on = publishDiagnosticOn;
-            expose_as_code_action = exposeAsCodeAction;
-            tsserver_path = tsserverPath;
-            tsserver_plugins = tsserverPlugins;
-            tsserver_max_memory = tsserverMaxMemory;
-            tsserver_format_options = tsserverFormatOptions;
-            tsserver_file_preferences = tsserverFilePreferences;
-            tsserver_locale = tsserverLocale;
-            complete_function_calls = completeFunctionCalls;
-            include_completions_with_insert_text = includeCompletionsWithInsertText;
-            code_lens = codeLens;
-            disable_member_code_lens = disableMemberCodeLens;
-            jsx_close_tag = with jsxCloseTag; {
-              inherit enable filetypes;
-            };
+
+  # NOTE: call setup in lsp config
+  callSetup = false;
+  extraConfig =
+    cfg:
+    let
+      definedOpts = lib.filter (opt: lib.hasAttrByPath opt cfg.settings) [
+        [ "separateDiagnosticServer" ]
+        [ "publishDiagnosticOn" ]
+        [ "exposeAsCodeAction" ]
+        [ "tsserverPath" ]
+        [ "tsserverPlugins" ]
+        [ "tsserverMaxMemory" ]
+        [ "tsserverFormatOptions" ]
+        [ "tsserverFilePreferences" ]
+        [ "tsserverLocale" ]
+        [ "completeFunctionCalls" ]
+        [ "includeCompletionsWithInsertText" ]
+        [ "codeLens" ]
+        [ "disableMemberCodeLens" ]
+        [
+          "jsxCloseTag"
+          "enable"
+        ]
+        [
+          "jsxCloseTag"
+          "filetypes"
+        ]
```

Comment:
Given these are the only nested options moved to settings.settings, and they only need snakeCase\_conversion at the top-level, we could dramatically simplify everything below by only handling the top-level attr rename?

So instead of a list of attr-paths, we'd have a list of (top-level) attr names.

Although, that may result in slightly less nice warning messages.

Reply:
Maybe we could keep the nested attrs for the warnings and "has attr" filtering, but then use `map head` to get the top-level attrs to simplify the renames?